src/server/game/Entities/Player/Player.cpp

```diff
@@ -27553,3 +27605,228 @@ void Player::SummonPet(uint32 entry, float x, float y, float z, float ang, PetTy
     AsynchPetSummon* asynchPetInfo = new AsynchPetSummon(entry, pos, petType, duration, createdBySpell, casterGUID);
     Pet::LoadPetFromDB(this, asynchLoadType, entry, 0, false, asynchPetInfo);
 }
+
+void Player::SetUnderACKmount()
+{
+    m_mountTimer = 3000;
+    m_ACKmounted = true;
+}
+
+void Player::SetRootACKUpd(uint32 delay)
+{
+    m_rootUpdTimer = 1500 + delay;
+    m_rootUpd = true;
+}
+
+bool Player::CheckOnFlyHack()
+{
+    if (IsCanFlybyServer())
+        return true;
+
+    if (ToUnit()->IsFalling() || IsFalling())
+        return true;
+
+    if (m_mover != this)
+        return true;
+
+    if (GetMapId() == 616 || GetMapId() == 649)
+        return true;
+
+    if (IsFlying() && !CanFly()) // kick flyhacks
+    {
+        sLog->outString("Player::CheckOnFlyHack :  FlyHack Detected for Account id : %u, Player %s", GetSession()->GetAccountId(), GetName().c_str());
+        sLog->outString("Player::========================================================");
+        sLog->outString("Player IsFlying but CanFly is false");
+
+        sWorld->SendGMText(LANG_GM_ANNOUNCE_AFH_CANFLYWRONG, GetName().c_str());
+        return false;
+    }
+
+    if (IsFlying() || IsLevitating() || IsInFlight())
+        return true;    
+
+    if (GetTransport() || GetVehicle() || GetVehicleKit())
+        return true;
+
+    if (HasAuraType(SPELL_AURA_CONTROL_VEHICLE))
+        return true;
+
+    if (HasUnitMovementFlag(MOVEMENTFLAG_ONTRANSPORT))
+        return true;
+
+    if (HasUnitState(UNIT_STATE_IGNORE_ANTISPEEDHACK))
+        return true;
+
+    if (UnderACKmount())
+        return true;
+
+    if (IsSkipOnePacketForASH())
+        return true;
+
+    Position npos;
+    GetPosition(&npos);
+    float pz = npos.GetPositionZ();
+    if (!IsInWater() && HasUnitMovementFlag(MOVEMENTFLAG_SWIMMING))
+    {
+        float waterlevel = GetBaseMap()->GetWaterLevel(npos.GetPositionX(), npos.GetPositionY()); // water walking
+        bool hovergaura = HasAuraType(SPELL_AURA_WATER_WALK) || HasAuraType(SPELL_AURA_HOVER);
+        if (waterlevel && (pz - waterlevel) <= (hovergaura ? GetCollisionHeight(IsMounted()) + 2.5f : GetCollisionHeight(IsMounted()) + 1.5f))
```

Comment:
`waterlevel` is a float but is being used as a boolean here, did you mean `waterlevel > 0` in the first condition?